lib/grpnames.gd

```diff
@@ -119,10 +229,32 @@
 ##      This is done by merging the two sequences into
 ##   a single increasing sequence of pairs <c_i, which_i>
 ##   where which_i indicates where c_i is in the a-sequence
-##   and where it is in the b-sequence if any. The the
+##   and where it is in the b-sequence if any. Then the
 ##   linear algorithm above may be used.
 ##
-DeclareAttribute( "DirectFactorsOfGroup", IsGroup );
+DeclareGlobalFunction( "DirectFactorsOfGroupFromList",
+                        [ IsGroup, IsList, IsList ] );
+
+#############################################################################
+##
+#A  DirectFactorsOfGroupKN( <G> ) . . . . decomposition into a direct product
+##
+##  <ManSection>
+##  <Attr Name="DirectFactorsOfGroupKN" Arg="G"/>
+##
+##  <Description>
+##    A (sorted if possible) list of factors [<A>G1</A>, .., <A>Gr</A>] such
+##    that <A>G</A> = <A>G1</A> x .. x <A>Gr</A> and none of the <A>Gi</A>
+##    is a direct product.
+##    This is the implementation of the algorithm described in
+##    Neeraj Kayal and Timur Nezhmetdinov, Factoring Groups Efficiently,
+##    in International Colloquium on Automata, Languages and Programming
+##    (ICALP), Lecture Notes in Computer Science 5555, 585-596,
+##    Springer Verlag, Berlin Heidelberg 2009.
+##  </Description>
+##  </ManSection>
+##
+DeclareAttribute( "DirectFactorsOfGroupKN", IsGroup );
```

Comment:
DirectFactorsOfGroupKN is almost _exactly_ the encoding of the algorithm in the Kayal-Nezhmetdinov paper, which is a polynomial algorithm in the _size_of_the_group_. However, I found that some parts of this algorithm is slow. Therefore, into the main code DirectFactorsOfGroup, I only incorporated the faster stuff, together with some of my own heuristics about checking for particular properties. 

Nevertheless, I decided to keep the complete code for the KN algorithm, because a) I spent a lot of time on putting it together and b) I believe it can be enhanced, but I simply decided not to do it now. 

I guess we could add an option "KN", and if somebody calls DirectFactorsOfGroup with this option, then it would call the KN method instead of the main algorithm. 

Does this make sense?